ui/src/app/views/workspace-edit/component.tsx

```diff
@@ -0,0 +1,590 @@
+        fontSize: (props.largeHeader ? 20 : 16)}}>
+        {props.header}
+      </div>
+      {props.required && <div style={styles.requiredText}>
+        (Required)
+      </div>
+      }
+      {props.tooltip && <TooltipTrigger content={props.tooltip}>
+        <InfoIcon style={{...styles.infoIcon,  marginTop: '0.2rem'}}/>
+      </TooltipTrigger>
+      }
+    </div>
+    {props.subHeader && <div style={{...styles.header, color: '#4A4A4A', fontSize: 14}}>
+      {props.subHeader}
+    </div>
+    }
+    <div style={styles.text}>
+      {props.text}
+    </div>
+    <div>
+      {props.children}
+    </div>
+  </div>;
+};
+
+export const WorkspaceCategory = (props) => {
+  return <div style={...fp.merge(styles.categoryRow, props.style)}>
+    <input style={{height: '.66667rem', marginRight: '.31667rem'}} type='checkbox'
+           checked={!!props.value}
+           onChange={props.onChange ? (e => props.onChange(e.target.checked)) : undefined}/>
+    <div style={{display: 'flex', flexDirection: 'column', marginTop: '-0.2rem'}}>
+      <label style={styles.shortDescription}>
+        {props.item.shortDescription}
+      </label>
+      <div>
+        <label style={styles.longDescription}>
+          {props.item.longDescription}
+        </label>
+        {props.children}
+      </div>
+    </div>
+  </div>;
+};
+
+export enum WorkspaceEditMode { Create = 1, Edit = 2, Clone = 3 }
+
+
+export interface WorkspaceEditProps {
+  routeConfigData: any;
+  workspace: Workspace;
+}
+
+export interface WorkspaceEditState {
+  cdrVersionItems: Array<CdrVersion>;
+  disableButton: Boolean;
+  workspace: Workspace;
+  workspaceCreationConflictError: Boolean;
+  workspaceCreationError: Boolean;
+  cloneUserRole: boolean;
+  loading: boolean;
+}
+
+export const WorkspaceEdit = withRouteConfigData()(withCurrentWorkspace()(
+  class extends React.Component<WorkspaceEditProps, WorkspaceEditState> {
+
+    constructor(props: WorkspaceEditProps) {
+      super(props);
+      this.isEmpty.bind(this);
+      this.state = {
+        cdrVersionItems: [],
+        disableButton: true,
+        workspace: {
+          name: '',
+          description: '',
+          dataAccessLevel: DataAccessLevel.Registered,
+          cdrVersionId: '',
+          researchPurpose: {
+            diseaseFocusedResearch: false,
+            methodsDevelopment: false,
+            controlSet: false,
+            aggregateAnalysis: false,
+            ancestry: false,
+            commercialPurpose: false,
+            population: false,
+            reviewRequested: false,
+            containsUnderservedPopulation: false,
+            underservedPopulationDetails: []
+          }
+        },
+        workspaceCreationConflictError: false,
+        workspaceCreationError: false,
+        cloneUserRole: false,
+        loading: false
+      };
+    }
+
+    componentWillMount() {
+      this.updateWorkspace();
+      this.setCdrVersions();
+    }
+
+
+    updateWorkspace() {
```

Comment:
Nit: This method name makes me think we are issuing an update request. Maybe `initializeWorkspace?`? It looks like it is always called on `componentWillMount`? It may also be worth just inlining there.

Reply:
So since workspace in initialized in constructor and this is updating the initialized workspace as per the mode, i named it 'updateWorkspace'.

Also i find it neater to have it as a separate method than inline

Reply:
That is completely reasonable, and I also find true, although in general I tend to think of functions as reuseable, but the idea with this is that it is called only once? I think that's why my preference is for inline.